feature(erase): Implement Erase Operation (#2334)
IBM/FHIR: operation/fhir-operation-erase/README.md

# Operation: fhir-operation-erase 

This document outlines the design and acceptance criteria for the *fhir-operation-erase*. 

**API Path** 
- `/fhir-server/api/v4/[RESOURCE]/[ID]/$erase`
- `/fhir-server/api/v4/[RESOURCE]/$erase`

|PATH Name|Required|Value|
|------|-----|-----|
|`RESOURCE`|required|The Resource Type|

> |`ID`|optional|The Resource ID, which implies the parameter logicalId must be set in the parameters object|

Reviewer: ```suggestion
|`ID`|optional|The Resource ID. If specified, the parameter logicalId must not be set in the parameters object. If not specified, the parameter logicalId must be set in the parameters object.|
```

**Method**
- POST - We do not support **GET**.

**Parameters**

|Parameter Name|Type|Value|Description|
|------|-----|-----|-----|
|`reason`|required|string|the textual reason to add to the Audit Event. A maximum of a 1000 characters.|
|`patient`|optional|string|the patient id to add to the Audit Event. Required, if the resource is being erased is within the Patient Compartment|

> |`id`|optional|string|Required, if the logical ID when executing an erase at the Resource Type level, else it must not be included.|

Reviewer: ```suggestion
|`id`|optional|string|Required, if executing an erase at the Resource Type level, else it must not be included.|
```

|`version`|optional|integer|If included, the version of the resource is erased.|

These values are passed as a FHIR Parameters Resource.

**HTTP Response Codes**

- 200 OK - The resource or the resource-version is permanently deleted.
- 400 BAD_REQUEST - The request is malformed.
- 401 NOT_AUTHENTICATED - The users is not authenticated.
- 403 NOT_AUTHORIZED - The user is not an administrator in an authorized user's group.
- 404 NOT_FOUND - The resource is not found.
- 500 INTERNAL SERVER ERROR - Failed to process the request.

> **Examples**

Reviewer: Would be nice to have an example of a deletion at the Resource Type level, and also of a non-Patient deletion.

Author: Added

            "valueInteger": 3
        }
    ]
}
```

#### Response - All Done delete a specific version

Response Code: 200 
Response Body: Return Operation Outcome.

``` json
{
    "resourceType": "Parameters",
    "id": "38fa6c99-69e0-4e6f-a134-f2658deebe07",
    "parameter": [
        {
            "name": "resource",
            "valueString": "Patient/17957682e44-bdd38817-6048-4db2-9502-9cb81175119b/_history/1"
        },
        {
            "name": "partial",
            "valueBoolean": true
        },
        {
            "name": "total",
            "valueInteger": 1
        }
    ]
}
```

**Server Invariants**

1. For Resources where the latest resource is deleted, FHIR search cannot be used to identify Resources specific to the Patient.
2. For Resources updated inflight, the operation may fail due to locking, and the client is expected to retry.

**Implementation Considerations**

1. Resources which are created incorrectly, should be $erased, rather than updated. For instance, a resource incorrectly identified with MRN1 which should be MRN2. However, deleting the offending version is possible once the offending version is NOT the latest version.

> 2. When execution a version on the Instance, erase removes all versions of a Resource to maintain the integrity of the datastore. When executing on the Instance and the version, erase removes the specific version.

Reviewer: ```suggestion
2. When executing on the Instance, erase removes all versions of a Resource to maintain the integrity of the datastore. When executing on the Instance and the version, erase removes the specific version.
```

|`id`|optional|string|Required, if executing an erase at the Resource Type level, else it must not be included.|
|`version`|optional|integer|If included, the version of the resource is erased.|

These values are passed as a FHIR Parameters Resource.

**HTTP Response Codes**

- 200 OK - The resource or the resource-version is permanently deleted.
- 400 BAD_REQUEST - The request is malformed.
- 401 NOT_AUTHENTICATED - The users is not authenticated.
- 403 NOT_AUTHORIZED - The user is not an administrator in an authorized user's group.
- 404 NOT_FOUND - The resource is not found.
- 500 INTERNAL SERVER ERROR - Failed to process the request.

**Examples**

### POST

*Delete all versions on Instance*

```
curl --location --request POST 'https://test.fhirexample.com/fhir-server/api/v4/Patient/1785fb0759b-1452d2e5-f568-442e-9841-3dc3940af5bc/$erase' \
--header 'Content-Type: application/fhir+json' \
--header 'Authorization: Basic ...' \
--data-raw '{
  "resourceType": "Parameters",
  "parameter": [
    {
      "name": "patient",
      "valueString": "patient-id-is-this-id"
    },
    {
      "name": "reason",
      "valueString": "My Reason for removing this resource"
    }
  ]
}'
```

> *Delete a specific version on Resource*

Reviewer: ```suggestion
*Delete all versions on Resource*
```
